# Khám phá dữ liệu

Notebook này để khám phá dữ liệu.

## Import những thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd

## Đọc dữ liệu từ file

File dữ liệu của đồ án sau bước thu thập dữ liệu đang nằm tại địa chỉ '../thu_thap_du_lieu/animal_crawler/animal_info.csv'. Ta sẽ tiến hành đọc dữ liệu đã thu thập và lưu vào DataFrame `animal_df`

In [3]:
animal_df = pd.read_csv('../thu_thap_du_lieu/animal_crawler/animals_info.csv')

Hiển thị 5 dòng đầu tiên trong dữ liệu để xem thử đã đọc đúng hay chưa

In [4]:
animal_df.head(5)

Name   Kingdom    Phylum   Subphylum     Class         Order  \
0     Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia     Carnivora   
1         Tiger  Animalia  Chordata  Vertebrata  Mammalia     Carnivora   
2    Brown Bear  Animalia  Chordata  Vertebrata  Mammalia     Carnivora   
3    Blue Whale  Animalia  Chordata  Vertebrata  Mammalia  Artiodactyla   
4  Killer Whale  Animalia  Chordata  Vertebrata  Mammalia  Artiodactyla   

     Suborder           Family         Genus                Species  ...  \
0  Caniformia          Canidae         Canis            Canis lupus  ...   
1  Feliformia          Felidae      Panthera        Panthera tigris  ...   
2  Caniformia          Ursidae         Ursus           Ursus arctos  ...   
3         NaN  Balaenopteridae  Balaenoptera  Balaenoptera musculus  ...   
4         NaN      Delphinidae       Orcinus           Orcinus orca  ...   

  Top speed      Weight     Height      Length  \
0   75 km/h    16-60 kg   80-85 cm  105-160 cm   
1   96 km/h   65-306 kg        NaN  200-390 cm   
2   56 km/h  100-635 kg  70-153 cm   1.4-2.8 m   
3   20 km/h   100-160 t        NaN     25-30 m   
4   45 km/h       3-6 t        NaN       6-9 m   

                                          Attributes  \
0  Nocturnal,Carnivore,Scavenger,Terrestrial,Curs...   
1  Nocturnal,Carnivore,Cursorial,Terrestrial,Ambu...   
2  Crepuscular,Omnivore,Terrestrial,Altricial,Nom...   
3  Carnivore,Piscivores,Aquatic,Precocial,Natator...   
4  Diurnal,Carnivore,Aquatic,Precocial,Cosmopolit...   

                                        Distribution  \
0  {'Geography': {'Continents': 'Asia, Europe, No...   
1  {'Geography': {'Continents': 'Asia', 'Subconti...   
2  {'Geography': {'Continents': 'Asia, Europe, No...   
3  {'Geography': {'Continents': 'Africa, South Am...   
4  {'Geography': {'Continents': 'Africa, Oceania,...   

                                              Habits                   Diet  \
0  {'Group name': 'pack, route, rout', 'Lifestyle...   Carnivore,,Scavenger   
1  {'Lifestyle': 'Cursorial, Terrestrial, Ambush ...              Carnivore   
2  {'Group name': 'sloth, sleuth', 'Lifestyle': '...               Omnivore   
3  {'Group name': 'pod, gam, herd, school, mod', ...  Carnivore,,Piscivores   
4  {'Group name': 'pod, gam, herd, school, mod', ...              Carnivore   

                                       Mating_Habits  \
0  {'Mating behavior': 'Monogamy', 'Reproduction ...   
1  {'Mating behavior': 'Polygyny', 'Reproduction ...   
2  {'Mating behavior': 'Polygynandry,', 'Reproduc...   
3  {'Mating behavior': 'Monogamy', 'Reproduction ...   
4  {'Mating behavior': 'Polygynandry', 'Reproduct...   

                                          Population  
0  {'Population trend': 'Stable', 'Population sta...  
1  {'Population trend': 'Decreasing', 'Population...  
2  {'Population trend': 'Stable', 'Population sta...  
3  {'Population trend': 'Increasing', 'Population...  
4       {'Population status': 'Data deficient (DD)'}  

[5 rows x 22 columns]

Vậy là đọc dữ liệu từ file đã ok

## Dữ liệu có bao nhiêu dòng, bao nhiêu cột?

In [5]:
animal_df.shape

(29357, 22)

## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng cho chúng ta biết các thông tin về một con vật, tất cả các dòng đều có cùng ý nghĩa như nhau.

## Dữ liệu có các dòng bị lặp không?

Chúng ta sử dụng hàm 'duplicated' trong pandas để kiểm tra dòng đó có phải bị lặp trong DataFrame hay không. Nếu kết quả là dòng đó bị lặp sẽ trả về giá trị 1, ngược lại sẽ là 0. Kế tiếp chúng ta sử dùng hàm 'sum' để tính tất cả giá trị trả về của hàm 'duplicated'. Nếu kết quả > 0 thì chứng tỏ có tồn tại dòng bị lặp. 

In [6]:
animal_df.duplicated().sum()

0

Vậy là dữ liệu của chúng ta không tồn lại dòng nào bị lặp lại.

## Mỗi cột có ý nghĩa gì?

In [7]:
animal_df.columns

Index(['Name', 'Kingdom', 'Phylum', 'Subphylum', 'Class', 'Order', 'Suborder',
       'Family', 'Genus', 'Species', 'Population size', 'Life span',
       'Top speed', 'Weight', 'Height', 'Length', 'Attributes', 'Distribution',
       'Habits', 'Diet', 'Mating_Habits', 'Population'],
      dtype='object')

## Dữ liệu lúc này có bao nhiêu dòng, bao nhiêu cột

In [8]:
animal_df.shape

(29357, 22)

## Các cột lúc này đang có kiểu dữ liệu gì

In [9]:
animal_df.dtypes

Name               object
Kingdom            object
Phylum             object
Subphylum          object
Class              object
Order              object
Suborder           object
Family             object
Genus              object
Species            object
Population size    object
Life span          object
Top speed          object
Weight             object
Height             object
Length             object
Attributes         object
Distribution       object
Habits             object
Diet               object
Mating_Habits      object
Population         object
dtype: object

## Tính xem tình trạng của mỗi cột dữ liệu hiện tại như thế nào

### Với mỗi cột ta sẽ tính:
- Tỉ lệ mất mát: missing_ratio
- Số lượng giá trị khác nhau: num_values
- Số lần xuất hiện của từng giá trị: value

Viết hàm "col_info" để tính và lưu kết quả vào mảng "cat_col_info_df"

In [10]:
def col_info(col, array):
    temp = pd.DataFrame(array[col])
    temp = temp.dropna()
    temp = np.array(temp[col])
    data = []
    # missing_ratio
    data.append(round((100 - (len(temp) / len(array[col]) * 100)), 1))
    # num_values
    temp = np.unique(temp, return_counts=True)
    data.append(len(temp[0]))
    # value
    temp = pd.Series(data = temp[1], index = temp[0])
    data.append(temp)
    # num_values
    return data

In [12]:
cat_col_info_df = pd.DataFrame(index = ['missing_ratio', 'num_values', 'value'])
for i in animal_df.columns:
    if animal_df[i].dtype not in [float, int]:
        cat_col_info_df[i] = col_info(i, animal_df)

cat_col_info_df

Name  \
missing_ratio                                                0.0   
num_values                                                 29357   
value          ? Nycticebus linglom               1
AEECL's s...   

                                                         Kingdom  \
missing_ratio                                                0.0   
num_values                                                   416   
value          Aaadonta constricta         2
Abert's Squirrel...   

                                                         Phylum  \
missing_ratio                                               2.8   
num_values                                                    2   
value          Chordata    23659
Mollusca     4866
dtype: int64   

                                                       Subphylum  \
missing_ratio                                               63.4   
num_values                                                     2   
value          Animalia        600
Vertebrata    10152
dtype:...   

                                                           Class  \
missing_ratio                                                0.0   
num_values                                                    12   
value          Amphibia           3393
Animalia            23...   

                                                           Order  \
missing_ratio                                                1.5   
num_values                                                   123   
value          Accipitriformes    255
Adapedonta          10
...   

                                                        Suborder  \
missing_ratio                                               89.9   
num_values                                                    16   
value          Alcedini            109
Caniformia          19...   

                                                          Family  \
missing_ratio                                                0.0   
num_values                                                   858   
value          Abrocomidae           10
Abronicidae          ...   

                                                           Genus  \
missing_ratio                                                0.0   
num_values                                                  6194   
value          Aaadonta        6
Abditomys       1
Abeillia  ...   

                                                         Species  ...  \
missing_ratio                                                0.0  ...   
num_values                                                 28688  ...   
value          "Calamita" melanorabdotus          1
"Calamita...  ...   

                                                       Top speed  \
missing_ratio                                               98.0   
num_values                                                   138   
value          0.24 km/h                   1
0.26 km/h       ...   

                                                          Weight  \
missing_ratio                                               90.2   
num_values                                                  2255   
value          0-0 kg            3
0.010-0.796 g     1
0.02-2...   

                                                          Height  \
missing_ratio                                               98.9   
num_values                                                   254   
value          0.5 m          1
0.7-1.3 m      1
0.75-1 m    ...   

                                                          Length  \
missing_ratio                                               90.0   
num_values                                                  2122   
value          0.5 m          2
0.5-1.8 m      1
0.6-2.43 m  ...   

                                                      Attributes  \
missing_ratio                                                0.0   
num_values                                                  3133   
value          Alg

Tại đây ta thấy các cột "Suborder", "Top speed", "Weight", "Height", "Length", "Diet" có missing_ratio rất cao (từ khoảng 90% trở lên)

Như vậy ta sẽ tiến hành xóa các dòng thiếu cả 3 giá trị 'Top speed', 'Weight', 'Length'

In [13]:
animal_df = animal_df.dropna(subset=['Top speed', 'Weight', 'Length'], how='all')
animal_df = animal_df.reset_index(drop = True)
# df.drop('index', inplace = True)
# df.head()

Vậy sau khi xóa đi các dòng như trên, tình trạng dữ liệu thay đổi như thế nào

In [15]:
cat_col_info_df = pd.DataFrame(index = ['missing_ratio', 'num_values', 'value'])
for i in animal_df.columns:
    if animal_df[i].dtype not in [float, int]:
        cat_col_info_df[i] = col_info(i, animal_df)

cat_col_info_df

Name  \
missing_ratio                                                0.0   
num_values                                                  3237   
value          Aardvark                      1
Aardwolf      ...   

                                                         Kingdom  \
missing_ratio                                                0.0   
num_values                                                    24   
value          African Wild Ass             1
American Crow  ...   

                                                       Phylum  \
missing_ratio                                             1.0   
num_values                                                  2   
value          Chordata    3135
Mollusca      69
dtype: int64   

                                                       Subphylum  \
missing_ratio                                               32.7   
num_values                                                     2   
value          Animalia        31
Vertebrata    2148
dtype: i...   

                                                           Class  \
missing_ratio                                                0.0   
num_values                                                     9   
value          Amphibia         79
Animalia          2
Aves  ...   

                                                           Order  \
missing_ratio                                                0.0   
num_values                                                    80   
value          Accipitriformes    67
Afrosoricida       25
An...   

                                                        Suborder  \
missing_ratio                                               76.1   
num_values                                                    15   
value          Alcedini             9
Caniformia         158
...   

                                                          Family  \
missing_ratio                                                0.0   
num_values                                                   387   
value          Abrocomidae        1
Accipitridae      60
Acro...   

                                                           Genus  \
missing_ratio                                                0.2   
num_values                                                  1622   
value          Abrawayaomys       1
Abrocoma           1
Acan...   

                                                         Species  ...  \
missing_ratio                                                0.0  ...   
num_values                                                  3218  ...   
value          Abrawayaomys ruschii            1
Abrocoma cin...  ...   

                                                       Top speed  \
missing_ratio                                               81.6   
num_values                                                   138   
value          0.24 km/h                   1
0.26 km/h       ...   

                                                          Weight  \
missing_ratio                                               11.2   
num_values                                                  2255   
value          0-0 kg            3
0.010-0.796 g     1
0.02-2...   

                                                          Height  \
missing_ratio                                               89.7   
num_values                                                   252   
value          0.5 m          1
0.7-1.3 m      1
0.75-1 m    ...   

                                                          Length  \
missing_ratio                                                9.0   
num_values                                                  2122   
value          0.5 m          2
0.5-1.8 m      1
0.6-2.43 m  ...   

                                                      Attributes  \
missing_ratio                                                0.0   
num_values                                                  2964   
value          Algivore,De

Ta thấy missing_ratio không thay đổi nhiều ở các cột 'Top speed', 'Length' nhưng lại giảm rất rõ rệt ở "Height" và "Diet"

## Xử lý tách các cột có kiểu dictionary thành nhiều cột

Đầu tiên các cột dictionary đều có kiểu là string nên ta phải xử lý chuyển các cột trên thành dict

In [19]:
for i in animal_df:
    if animal_df[i][0].find('{') != -1:
        animal_df[i] = animal_df[i].apply(eval)

Viết hàm "split_col" để split các cột có kiểu dữ liệu là dict thành nhiều cột

In [20]:
def split_col(df):
    element = []
    for i in df:
        if type(df[i][0]) == dict: 
            element.append(i)
    df_split_animals_info = pd.DataFrame()
    for i in df:
        if i not in element:
            df_split_animals_info[i] = df[i]
        else:
            temp3 = df[i][0].items()
            temp3 = np.array(list(temp3), dtype = object)
            df_temp2 = pd.DataFrame(columns = temp3[:, 0])
            for rows in df[i]:
                if type(rows) != float:
                    temp3 = rows.items()
                else:
                    temp3 = []
                temp3 = np.array(list(temp3), dtype = object)
                if len(temp3) > 1: 
                    df_temp = pd.DataFrame(index = temp3[:, 0], data = temp3[:, 1]).T
                else:
                    df_temp = pd.DataFrame(columns = df_temp2.columns, index = [0])
                df_temp2 = pd.concat([df_temp2, df_temp], ignore_index=True)
            df_split_animals_info = df_split_animals_info.join(df_temp2)
    return df_split_animals_info
    

In [21]:
df_split_animals_info = split_col(animal_df)
pd.set_option('display.max_columns', None)
df_split_animals_info

Name   Kingdom    Phylum   Subphylum     Class  \
0                    Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1                        Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2                   Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3                   Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4                 Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
...                        ...       ...       ...         ...       ...   
3232     Yellow-billed cacique  Animalia  Chordata         NaN      Aves   
3233        Saint Lucia oriole  Animalia  Chordata         NaN      Aves   
3234        Swynnerton's robin  Animalia  Chordata         NaN      Aves   
3235  Ochre-bellied flycatcher  Animalia  Chordata         NaN      Aves   
3236  Magellanic diving petrel  Animalia  Chordata         NaN      Aves   

                  Order    Suborder           Family         Genus  \
0             Carnivora  Caniformia          Canidae         Canis   
1             Carnivora  Feliformia          Felidae      Panthera   
2             Carnivora  Caniformia          Ursidae         Ursus   
3          Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4          Artiodactyla         NaN      Delphinidae       Orcinus   
...                 ...         ...              ...           ...   
3232      Passeriformes         NaN        Icteridae   Amblycercus   
3233      Passeriformes         NaN        Icteridae       Icterus   
3234      Passeriformes         NaN     Muscicapidae  Swynnertonia   
3235      Passeriformes         NaN       Tyrannidae     Mionectes   
3236  Procellariiformes         NaN   Procellariidae  Pelecanoides   

                       Species Population size   Life span Top speed  \
0                  Canis lupus         400,000   10-20 yrs   75 km/h   
1              Panthera tigris     2,154-3,159   10-15 yrs   96 km/h   
2                 Ursus arctos         200,000   20-50 yrs   56 km/h   
3        Balaenoptera musculus      10-25 Thou   80-90 yrs   20 km/h   
4                 Orcinus orca          50,000  30-100 yrs   45 km/h   
...                        ...             ...         ...       ...   
3232  Amblycercus holosericeus             NaN         NaN       NaN   
3233        Icterus laudabilis             NaN         NaN       NaN   
3234  Swynnertonia swynnertoni             NaN         NaN       NaN   
3235      Mionectes oleagineus             NaN         NaN       NaN   
3236    Pelecanoides magellani             NaN         NaN       NaN   

           Weight     Height        Length  \
0        16-60 kg   80-85 cm    105-160 cm   
1       65-306 kg        NaN    200-390 cm   
2      100-635 kg  70-153 cm     1.4-2.8 m   
3       100-160 t        NaN       25-30 m   
4           3-6 t        NaN         6-9 m   
...           ...        ...           ...   
3232  56.7-70.9 g        NaN  21.6-25.4 cm   
3233       36.9 g        NaN      20-22 cm   
3234          NaN        NaN      12-13 cm   
3235         13 g        NaN           NaN   
3236    145-174 g        NaN         19 cm   

                                             Attributes  \
0     Nocturnal,Carnivore,Scavenger,Terrestrial,Curs...   
1     Nocturnal,Carnivore,Cursorial,Terrestrial,Ambu...   
2     Crepuscular,Omnivore,Terrestrial,Altricial,Nom...   
3     Carnivore,Piscivores,Aquatic,Precocial,Natator...   
4     Diurnal,Carnivore,Aquatic,Precocial,Cosmopolit...   
...                                                 ...   
3232  Diurnal,Omnivore,Zoochory,Terrestrial,Monogamy...   
3233  Diurnal,Herbivore,Frugivore,Nectarivore,Arbore...   
3234  Diurnal,Carnivore,Insectivores,Terrestrial,Not...   
3235              Terrestrial,Not a migrant,starts with   
3236  Diurnal,Nocturnal,Carnivore,Piscivores,Terrest...   

                                              Geography  \
0     {'Continents': 'Asia, Europe, North America, A...   
1     {'Continents': 'Asia'

Sau khi split dữ liệu có số dòng là 3237 và số cột là 37, tăng lên 15 cột

Nhưng nhìn lại dữ liệu vẫn còn cột là dictionary như "Geograpy", như vậy ta phải split thêm 1 lần nữa

In [22]:
df_split_animals_info = split_col(df_split_animals_info)
df_split_animals_info

Name   Kingdom    Phylum   Subphylum     Class  \
0                    Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1                        Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2                   Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3                   Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4                 Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
...                        ...       ...       ...         ...       ...   
3232     Yellow-billed cacique  Animalia  Chordata         NaN      Aves   
3233        Saint Lucia oriole  Animalia  Chordata         NaN      Aves   
3234        Swynnerton's robin  Animalia  Chordata         NaN      Aves   
3235  Ochre-bellied flycatcher  Animalia  Chordata         NaN      Aves   
3236  Magellanic diving petrel  Animalia  Chordata         NaN      Aves   

                  Order    Suborder           Family         Genus  \
0             Carnivora  Caniformia          Canidae         Canis   
1             Carnivora  Feliformia          Felidae      Panthera   
2             Carnivora  Caniformia          Ursidae         Ursus   
3          Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4          Artiodactyla         NaN      Delphinidae       Orcinus   
...                 ...         ...              ...           ...   
3232      Passeriformes         NaN        Icteridae   Amblycercus   
3233      Passeriformes         NaN        Icteridae       Icterus   
3234      Passeriformes         NaN     Muscicapidae  Swynnertonia   
3235      Passeriformes         NaN       Tyrannidae     Mionectes   
3236  Procellariiformes         NaN   Procellariidae  Pelecanoides   

                       Species Population size   Life span Top speed  \
0                  Canis lupus         400,000   10-20 yrs   75 km/h   
1              Panthera tigris     2,154-3,159   10-15 yrs   96 km/h   
2                 Ursus arctos         200,000   20-50 yrs   56 km/h   
3        Balaenoptera musculus      10-25 Thou   80-90 yrs   20 km/h   
4                 Orcinus orca          50,000  30-100 yrs   45 km/h   
...                        ...             ...         ...       ...   
3232  Amblycercus holosericeus             NaN         NaN       NaN   
3233        Icterus laudabilis             NaN         NaN       NaN   
3234  Swynnertonia swynnertoni             NaN         NaN       NaN   
3235      Mionectes oleagineus             NaN         NaN       NaN   
3236    Pelecanoides magellani             NaN         NaN       NaN   

           Weight     Height        Length  \
0        16-60 kg   80-85 cm    105-160 cm   
1       65-306 kg        NaN    200-390 cm   
2      100-635 kg  70-153 cm     1.4-2.8 m   
3       100-160 t        NaN       25-30 m   
4           3-6 t        NaN         6-9 m   
...           ...        ...           ...   
3232  56.7-70.9 g        NaN  21.6-25.4 cm   
3233       36.9 g        NaN      20-22 cm   
3234          NaN        NaN      12-13 cm   
3235         13 g        NaN           NaN   
3236    145-174 g        NaN         19 cm   

                                             Attributes  \
0     Nocturnal,Carnivore,Scavenger,Terrestrial,Curs...   
1     Nocturnal,Carnivore,Cursorial,Terrestrial,Ambu...   
2     Crepuscular,Omnivore,Terrestrial,Altricial,Nom...   
3     Carnivore,Piscivores,Aquatic,Precocial,Natator...   
4     Diurnal,Carnivore,Aquatic,Precocial,Cosmopolit...   
...                                                 ...   
3232  Diurnal,Omnivore,Zoochory,Terrestrial,Monogamy...   
3233  Diurnal,Herbivore,Frugivore,Nectarivore,Arbore...   
3234  Diurnal,Carnivore,Insectivores,Terrestrial,Not...   
3235              Terrestrial,Not a migrant,starts with   
3236  Diurnal,Nocturnal,Carnivore,Piscivores,Terrest...   

                                             Continents  \
0                   Asia, Europe, North America, Africa   
1                          

Sau khi split dữ liệu tăng thêm 9 cột

Lúc này ta tính lại missing_ratio xem các cột sau khi split như thế nào nhé

In [23]:
cat_col_info_df = pd.DataFrame(index = ['missing_ratio', 'num_values', 'value'])
for i in df_split_animals_info.columns:
    if df_split_animals_info[i].dtype not in [float, int]:
        cat_col_info_df[i] = col_info(i, df_split_animals_info)

cat_col_info_df

Name  \
missing_ratio                                                0.0   
num_values                                                  3237   
value          Aardvark                      1
Aardwolf      ...   

                                                         Kingdom  \
missing_ratio                                                0.0   
num_values                                                    24   
value          African Wild Ass             1
American Crow  ...   

                                                       Phylum  \
missing_ratio                                             1.0   
num_values                                                  2   
value          Chordata    3135
Mollusca      69
dtype: int64   

                                                       Subphylum  \
missing_ratio                                               32.7   
num_values                                                     2   
value          Animalia        31
Vertebrata    2148
dtype: i...   

                                                           Class  \
missing_ratio                                                0.0   
num_values                                                     9   
value          Amphibia         79
Animalia          2
Aves  ...   

                                                           Order  \
missing_ratio                                                0.0   
num_values                                                    80   
value          Accipitriformes    67
Afrosoricida       25
An...   

                                                        Suborder  \
missing_ratio                                               76.1   
num_values                                                    15   
value          Alcedini             9
Caniformia         158
...   

                                                          Family  \
missing_ratio                                                0.0   
num_values                                                   387   
value          Abrocomidae        1
Accipitridae      60
Acro...   

                                                           Genus  \
missing_ratio                                                0.2   
num_values                                                  1622   
value          Abrawayaomys       1
Abrocoma           1
Acan...   

                                                         Species  \
missing_ratio                                                0.0   
num_values                                                  3218   
value          Abrawayaomys ruschii            1
Abrocoma cin...   

                                                 Population size  \
missing_ratio                                               44.4   
num_values                                                   652   
value          0.45-1 M        1
1 Mln           2
1,000     ...   

                                                       Life span  \
missing_ratio                                               42.9   
num_values                                                   603   
value          1 yr            2
1-10 yrs        2
1-12.5 yrs...   

                                                       Top speed  \
missing_ratio                                               81.6   
num_values                                                   138   
value          0.24 km/h                   1
0.26 km/h       ...   

                                                          Weight  \
missing_ratio                                               11.2   
num_values                                                  2255   
value          0-0 kg            3
0.010-0.796 g     1
0.02-2...   

                                                          Height  \
missing_ratio                                               89.7   
num_values                                                   252   
value          0.5 m          1
0.7-1.3 m     

## Tiếp tục tiền xử lý dữ liệu nhé

Tại đây ta sẽ tách các cột có thuộc tính cách nhau bằng dấu phẩy thành 1 list

Luu vào "element" là tên các cột có các thuộc tính phân tách bằng dấu phẩy

Các cột sau khi xử lý sẽ được lưu vào biến "splits"

In [24]:
element = np.array(['Attributes', 'Continents', 'Subcontinents', 'Countries', 'Regions', 'Wwf biomes', 'Group name', 'Diet', 'Biogeographical realms', 'Lifestyle', 'Oceans'])

splits = [df_split_animals_info[col].str.split(pat = ',', expand = False) for col in element]
splits = pd.DataFrame(splits).T
# print(splits.shape)
# print(splits.head())


Sau khi tách xong ta tiến hành ghép 2 dataframe lại thành "data_df"

In [25]:
df_split_animals_info.drop(element, axis=1, inplace=True)
data_df = pd.concat([df_split_animals_info, splits], axis = 1)
# data_df = data_df.drop(columns = data_df.columns[0], axis = 1)
data_df
# df_split_animals_info

Name   Kingdom    Phylum   Subphylum     Class  \
0                    Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1                        Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2                   Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3                   Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4                 Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
...                        ...       ...       ...         ...       ...   
3232     Yellow-billed cacique  Animalia  Chordata         NaN      Aves   
3233        Saint Lucia oriole  Animalia  Chordata         NaN      Aves   
3234        Swynnerton's robin  Animalia  Chordata         NaN      Aves   
3235  Ochre-bellied flycatcher  Animalia  Chordata         NaN      Aves   
3236  Magellanic diving petrel  Animalia  Chordata         NaN      Aves   

                  Order    Suborder           Family         Genus  \
0             Carnivora  Caniformia          Canidae         Canis   
1             Carnivora  Feliformia          Felidae      Panthera   
2             Carnivora  Caniformia          Ursidae         Ursus   
3          Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4          Artiodactyla         NaN      Delphinidae       Orcinus   
...                 ...         ...              ...           ...   
3232      Passeriformes         NaN        Icteridae   Amblycercus   
3233      Passeriformes         NaN        Icteridae       Icterus   
3234      Passeriformes         NaN     Muscicapidae  Swynnertonia   
3235      Passeriformes         NaN       Tyrannidae     Mionectes   
3236  Procellariiformes         NaN   Procellariidae  Pelecanoides   

                       Species Population size   Life span Top speed  \
0                  Canis lupus         400,000   10-20 yrs   75 km/h   
1              Panthera tigris     2,154-3,159   10-15 yrs   96 km/h   
2                 Ursus arctos         200,000   20-50 yrs   56 km/h   
3        Balaenoptera musculus      10-25 Thou   80-90 yrs   20 km/h   
4                 Orcinus orca          50,000  30-100 yrs   45 km/h   
...                        ...             ...         ...       ...   
3232  Amblycercus holosericeus             NaN         NaN       NaN   
3233        Icterus laudabilis             NaN         NaN       NaN   
3234  Swynnertonia swynnertoni             NaN         NaN       NaN   
3235      Mionectes oleagineus             NaN         NaN       NaN   
3236    Pelecanoides magellani             NaN         NaN       NaN   

           Weight     Height        Length Introduced countries Seas Islands  \
0        16-60 kg   80-85 cm    105-160 cm                  NaN  NaN     NaN   
1       65-306 kg        NaN    200-390 cm                  NaN  NaN     NaN   
2      100-635 kg  70-153 cm     1.4-2.8 m                  NaN  NaN     NaN   
3       100-160 t        NaN       25-30 m                  NaN  NaN     NaN   
4           3-6 t        NaN         6-9 m                  NaN  NaN     NaN   
...           ...        ...           ...                  ...  ...     ...   
3232  56.7-70.9 g        NaN  21.6-25.4 cm                  NaN  NaN     NaN   
3233       36.9 g        NaN      20-22 cm                  NaN  NaN     NaN   
3234          NaN        NaN      12-13 cm                  NaN  NaN     NaN   
3235         13 g        NaN           NaN                  NaN  NaN     NaN   
3236    145-174 g        NaN         19 cm                  NaN  NaN     NaN   

                                                  Biome  \
0     [Wetlands , Shrubland, Forest, Mountains, Anth...   
1     [Forest, Rainforest, Montane forest, Riparian,...   
2                                    [Forest, Woodland]   
3          [Oceanic pelagic zone, Marine, Neritic zone]   
4     [Oceanic pelagic zone, Marine, Neritic zone, E...   
...                                                 ...   
3232                             

Xử lý tại các cột list, những giá trị nan sẽ được thay thế thành một list rỗng []

In [26]:
data_df[element] = data_df[element].fillna(data_df.notna().applymap(lambda x: x or []))
# data_df = data_df.drop(columns = data_df.columns[0], axis = 1)
data_df

Name   Kingdom    Phylum   Subphylum     Class  \
0                    Grey Wolf  Animalia  Chordata  Vertebrata  Mammalia   
1                        Tiger  Animalia  Chordata  Vertebrata  Mammalia   
2                   Brown Bear  Animalia  Chordata  Vertebrata  Mammalia   
3                   Blue Whale  Animalia  Chordata  Vertebrata  Mammalia   
4                 Killer Whale  Animalia  Chordata  Vertebrata  Mammalia   
...                        ...       ...       ...         ...       ...   
3232     Yellow-billed cacique  Animalia  Chordata         NaN      Aves   
3233        Saint Lucia oriole  Animalia  Chordata         NaN      Aves   
3234        Swynnerton's robin  Animalia  Chordata         NaN      Aves   
3235  Ochre-bellied flycatcher  Animalia  Chordata         NaN      Aves   
3236  Magellanic diving petrel  Animalia  Chordata         NaN      Aves   

                  Order    Suborder           Family         Genus  \
0             Carnivora  Caniformia          Canidae         Canis   
1             Carnivora  Feliformia          Felidae      Panthera   
2             Carnivora  Caniformia          Ursidae         Ursus   
3          Artiodactyla         NaN  Balaenopteridae  Balaenoptera   
4          Artiodactyla         NaN      Delphinidae       Orcinus   
...                 ...         ...              ...           ...   
3232      Passeriformes         NaN        Icteridae   Amblycercus   
3233      Passeriformes         NaN        Icteridae       Icterus   
3234      Passeriformes         NaN     Muscicapidae  Swynnertonia   
3235      Passeriformes         NaN       Tyrannidae     Mionectes   
3236  Procellariiformes         NaN   Procellariidae  Pelecanoides   

                       Species Population size   Life span Top speed  \
0                  Canis lupus         400,000   10-20 yrs   75 km/h   
1              Panthera tigris     2,154-3,159   10-15 yrs   96 km/h   
2                 Ursus arctos         200,000   20-50 yrs   56 km/h   
3        Balaenoptera musculus      10-25 Thou   80-90 yrs   20 km/h   
4                 Orcinus orca          50,000  30-100 yrs   45 km/h   
...                        ...             ...         ...       ...   
3232  Amblycercus holosericeus             NaN         NaN       NaN   
3233        Icterus laudabilis             NaN         NaN       NaN   
3234  Swynnertonia swynnertoni             NaN         NaN       NaN   
3235      Mionectes oleagineus             NaN         NaN       NaN   
3236    Pelecanoides magellani             NaN         NaN       NaN   

           Weight     Height        Length Introduced countries Seas Islands  \
0        16-60 kg   80-85 cm    105-160 cm                  NaN  NaN     NaN   
1       65-306 kg        NaN    200-390 cm                  NaN  NaN     NaN   
2      100-635 kg  70-153 cm     1.4-2.8 m                  NaN  NaN     NaN   
3       100-160 t        NaN       25-30 m                  NaN  NaN     NaN   
4           3-6 t        NaN         6-9 m                  NaN  NaN     NaN   
...           ...        ...           ...                  ...  ...     ...   
3232  56.7-70.9 g        NaN  21.6-25.4 cm                  NaN  NaN     NaN   
3233       36.9 g        NaN      20-22 cm                  NaN  NaN     NaN   
3234          NaN        NaN      12-13 cm                  NaN  NaN     NaN   
3235         13 g        NaN           NaN                  NaN  NaN     NaN   
3236    145-174 g        NaN         19 cm                  NaN  NaN     NaN   

                                                  Biome  \
0     [Wetlands , Shrubland, Forest, Mountains, Anth...   
1     [Forest, Rainforest, Montane forest, Riparian,...   
2                                    [Forest, Woodland]   
3          [Oceanic pelagic zone, Marine, Neritic zone]   
4     [Oceanic pelagic zone, Marine, Neritic zone, E...   
...                                                 ...   
3232                             

## Thống kê các chỉ số cho các cột có kiểu dữ liệu là list

In [27]:
def stat_list_col(col, array):
    temp = pd.DataFrame(array[col])
    temp = temp.dropna()
    temp = np.array(temp[col])
    data = []
    # missing_ratio
    data.append(round((100 - (len(temp) / len(array[col]) * 100)), 1))
    # num_values
    temp = np.hstack(temp)
    temp = list(map(str.strip, temp)) # Xử lý các dấu cách đầu và cuối mỗi chuỗi
    temp = np.unique(temp, return_counts=True)
    data.append(len(temp[0]))
    # value
    temp = pd.Series(data = temp[1], index = temp[0])
    data.append(temp)
    # num_values
    return data

#     total = data_df[col].tolist()
#     total = np.hstack(total)
#     value, times = np.unique(total, return_counts = True)
    

In [28]:
list_col_info_df = pd.DataFrame(index = ['missing_ratio', 'num_values', 'value'])
for i in data_df:
    if type(data_df[i][0]) == list:
        list_col_info_df[i] = stat_list_col(i, data_df)
list_col_info_df

Biome  \
missing_ratio                                                0.0   
num_values                                                    43   
value          Abyssal zone               4
Agricultural     ...   

                                                         Climate  \
missing_ratio                                                0.0   
num_values                                                     6   
value          Arid          304
Cold          309
Libya     ...   

                                                      Attributes  \
missing_ratio                                                0.0   
num_values                                                   139   
value          Aestivation              71
Algivore          ...   

                                                      Continents  \
missing_ratio                                                0.0   
num_values                                                    76   
value                                  3
Africa              ...   

                                                   Subcontinents  \
missing_ratio                                                0.0   
num_values                                                    89   
value                                       12
Afghanistan   ...   

                                                       Countries  \
missing_ratio                                                0.0   
num_values                                                   251   
value                            884
Afghanistan        66
Af...   

                                                         Regions  \
missing_ratio                                                0.0   
num_values                                                   394   
value                          65
Acre             2
Afghanis...   

                                                      Wwf biomes  \
missing_ratio                                                0.0   
num_values                                                   249   
value                                                        ...   

                                                      Group name  \
missing_ratio                                                0.0   
num_values                                                   402   
value                            1
"bind"            1
"littl...   

                                                            Diet  \
missing_ratio                                                0.0   
num_values                                                    30   
value                            2941
Algivore            16
...   

                                          Biogeographical realms  \
missing_ratio                                                0.0   
num_values                                                   391   
value                                                       3...   

                                                       Lifestyle  \
missing_ratio                                                0.0   
num_values                                                    51   
value                                              1784
Aesti...   

                                                          Oceans  
missing_ratio                                                0.0  
num_values                                                   253  
value                               10
Afghanistan           ...

### Lưu dữ liệu đã tiền xử lý vào file 'animals_split_info.csv' để tiếp tục sử dụng trong các phần sau

In [29]:
data_df.to_csv('animals_split_info.csv')